Create readability score aggregator

hansjovis opened the pull request:
## Summary

This PR can be summarized in the following changelog entry:

* _[non-user-facing]_ Adds a `ReadabilityScoreAggregator` that should serve as a replacement for the score aggregation part of the current `contentAssessor` in the (near) future.

## Relevant technical choices:

* Uses the same logic as in `contentAssessor`.
* May get some merge conflicts when #2173 is merged first, since it restructures the folder structure somewhat. These changes may not coincide with the structure in this branch.

## Test instructions

This PR can be tested by following these steps:

* Look at the code.
    * Specifically, check if the logic used is the same as the score aggregation functionality in `contentAssessor` (The `aggregate` method in the `ReadabilityScoreAggregator` method should implement the same logic as the `calculateOverallScore` method).
 * Run the tests.
    * Check that the coverage is 100% for the `ReadabilityScoreAggregator`.
    * Check that all tests pass.

Fixes #2125

hansjovis committed:
Adds basic structure for ReadabilityScoreAggregator class.

hansjovis committed:
Implemented first version of ReadabilityScoreAggregator.

hansjovis committed:
Added basic test case.

hansjovis committed:
Added tests.

hansjovis committed:
Adds JSDoc comment to getValidResults.

hansjovis committed:
Adds an extra test to get coverage to 100%.

igorschoester commented on src/tree/assess/scoreAggregators/ReadabilityScoreAggregator.js:
```diff
@@ -0,0 +1,167 @@
+	 * @returns {number} The total penalty points for the results.
+	 */
+	calculatePenalty( results ) {
+		const penaltyPoints = results.map( result => {
```
The usage of the `reduce` function here would combine the `map` and `sum` 😉

hansjovis replied:
Ah, nice one! 
I reduced the `map` and `sum` calls into one `reduce` call.

igorschoester commented on src/tree/assess/scoreAggregators/ReadabilityScoreAggregator.js:
```diff
@@ -0,0 +1,167 @@
+	 *
+	 * @returns {number} The total penalty points for the results.
+	 */
+	calculatePenalty( results ) {
```
Is it `penalty` or `penaltyPoints`? I actually like your decision to move away from the `points` part. However, you do not use that consistently. Below they are `penaltyPoints` again, while in `calculateScore` you pass `penalty`.

hansjovis replied:
The code and comments now always refer to `penalty` 🙂instead of `penaltyPoints`.

igorschoester commented on src/tree/assess/scoreAggregators/ReadabilityScoreAggregator.js:
```diff
@@ -0,0 +1,167 @@
+			// Compute the rating ("error", "feedback", "bad", "ok" or "good").
+			const rating = scoreToRating( result.getScore() );
+
+			if ( this.allAssessmentsSupported( results ) ) {
```
What do you think of this snippet?
```js
const penalty = this.allAssessmentsSupported( results )
	? penaltyMappingFullSupport[ rating ]
	: penaltyMappingPartialSupport[ rating ];
return penalty || 0;
```
I think a ternary operator here makes it more readable. Plus the `|| 0` fallback is only needed once.

hansjovis replied:
Changed the code to use the ternary operator.

igorschoester commented on src/tree/assess/scoreAggregators/ReadabilityScoreAggregator.js:
```diff
@@ -0,0 +1,167 @@
+ * @type {number}
+ * @const
+ */
+const totalNrOfAssessments = 8;
```
Do we have to keep this number updated? Maybe an improvement on this would be nicer. I'm thinking a key count of the readability index? Similar to how we do this for language functionality (`getFunctionWordsLanguages`).

hansjovis replied:
Yes. I think the most straight forward way is the way that you describe.
So: a file with all the readability assessments that gets
1. Imported (or defined) and used in the to-be-implemented `constructReadabilityAssessor()` function to make a new `TreeAssessor` for the readability analysis.
2. Gets imported here, with this line changing to (although it can be wrapped into a helper function, if needed):
  ```js
const totalNrOfAssessments = readabilityAssessments.length;
  ```

igorschoester commented on src/tree/assess/scoreAggregators/ReadabilityScoreAggregator.js:
```diff
@@ -0,0 +1,167 @@
+	 *
+	 * @returns {boolean} True if fully supported.
+	 */
+	allAssessmentsSupported( results ) {
```
I hope this name can be improved on. Maybe `hasFullSupport` or `isFullySupported`, to keep in the same line as the `penaltyMapping` constants.

hansjovis replied:
I like `isFullySupported`. Changed the method name!

igorschoester commented on src/tree/assess/scoreAggregators/ReadabilityScoreAggregator.js:
```diff
@@ -0,0 +1,167 @@
+
+		/*
+		 * If you have no content, you have a red indicator.
+		 * (Assume that one result always means the 'no content' assessment result, (see `contentAssessor`) ).
```
Let's not refer to soon-to-be-deprecated code in the comments?

hansjovis replied:
You're right. I updated the comment to not refer to the soon-to-be-deprecated class.

igorschoester commented on src/tree/assess/scoreAggregators/ReadabilityScoreAggregator.js:
```diff
@@ -0,0 +1,167 @@
+			return READABILITY_SCORES.NEEDS_IMPROVEMENT;
+		}
+
+		const penalty = this.calculatePenalty( results );
```
Shouldn't this be calculated from the valid results?

hansjovis replied:
Good catch. Changed it to `validResults`.

igorschoester commented on src/tree/assess/scoreAggregators/ReadabilityScoreAggregator.js:
```diff
@@ -0,0 +1,167 @@
+		}
+
+		const penalty = this.calculatePenalty( results );
+		return this.calculateScore( this.allAssessmentsSupported( results ), penalty );
```
Since you saved the penalty on a variable, it seems nice to do the same with the supported variable.
Also, doesn't `results` need to be valid results?

hansjovis replied:
Saved the result of `isFullySupported` to a variable, which I now give as an argument to the function.

The `contentAssessor` checks if the number of applicable assessments is the same as the total number of available assessments. It does not check for validity, since it does not check the assessment results. Either way, we need to change the way language support is checked, since it does not work the way it says it does now.
An assessment can be applicable because of a lot of different reasons, not only because the language is supported.

We could use this function:
```js
const isLanguageAvailable = getLanguageAvailability( paper.getLocale(), availableLanguages );
```
Although we then need a way to get the locale of the paper within the `Aggregator`.

igorschoester commented on spec/tree/assess/scoreAggregators/ReadabilityScoreAggregatorSpec.js:
```diff
@@ -0,0 +1,156 @@
+		let aggregator;
+
+		beforeEach( function() {
+			aggregator = new ReadabilityScoreAggregator();
```
I should do it like this in the `SEOScoreAggregator` specs too.

igorschoester reviewed (changes requested):
CR 🏗 

Some questions and improvement suggestions.

hansjovis committed:
Refactored `calculatePenalty`.

Made use of `reduce` instead of map and sum; used ternary operator when computing penalty points.

hansjovis committed:
Comments in ReadabilityScoreAggregator now always refer to 'penalty' instead of 'penalty points'.

hansjovis committed:
Added todo comment to make sure we compute the total number of readability assessments, instead of setting it as a constant.

hansjovis committed:
Changed  to .

hansjovis committed:
Penalty scores are now calculated based on the valid results, rather than all the results.

hansjovis committed:
Changed call to isFullySupported in argument list to a variable which I give as an argument instead.

igorschoester commented on src/tree/assess/scoreAggregators/ReadabilityScoreAggregator.js:
```diff
@@ -0,0 +1,169 @@
+/**
+ * Total number of available readability assessments.
+ *
+ * To do: Compute this, rather than set this and keep this updated (see https://github.com/Yoast/YoastSEO.js/pull/2176#discussion_r261547836)
```
Please create an issue for this instead of a comment in the code!

As discussed, a second look at how we want to score readability would be nice. The logic in this is now a copy of the old one. It is pretty complex and I have the feeling it could be improved. However, it would be better to do that in an iteration.

This could be done in cooperation with someone that has the knowledge on why this is the way it is.

hansjovis replied:
I agree. It can be improved and we definitely need to ask someone who knows the ins- and outs of the current scoring mechanism.

igorschoester commented on src/tree/assess/scoreAggregators/ReadabilityScoreAggregator.js:
```diff
@@ -0,0 +1,169 @@
+	 * Calculates the total score (GOOD, OKAY or NEEDS IMPROVEMENT)
+	 * based on the total penalty score.
+	 *
+	 * @param {boolean} allAssessmentsSupported Whether all assessments are supported for this language.
```
I think you forgot to rename this one to `isFullySupported` 😉

igorschoester commented on src/tree/assess/scoreAggregators/ReadabilityScoreAggregator.js:
```diff
@@ -0,0 +1,169 @@
+	 */
+	calculateScore( allAssessmentsSupported, penalty ) {
+		if ( allAssessmentsSupported ) {
+			if ( penalty > 6 ) {
```
Magic numbers: `6`, `4` and `4`, `2`. Let me try to understand.

`isFullySupported`: When the amount of results is the same as the expected amount.
`penalty`: The sum of penalty points for each result.

When the penalty is higher than `6`, the readability needs improvement.
When the penalty is higher than `4`, the readability is ok.
When the penalty is `4` or lower, the readability is good.

To make up for the lesser amount of results, the penalty thresholds are lower for when it is not fully supported.
Meaning that a lower amount of gathered penalties will result in worse scores sooner.

This is on top of the already higher penalty point being given to the `bad` score of the partially supported. Which might just be done for finer control, maybe a better break between the `bad` and `ok` scores. I'm just guessing here. Maybe @IreneStr can enlighten us a bit here?

igorschoester replied:
In general, I would hope some more comments here to explain the magic numbers and proces. But I understand you are also just guessing 😅

hansjovis replied:
It is actually quite simple once you look at the penalty scores given to each result type (`good`, `ok` or `bad`). E.g. two `bad` results give a total score of 6 (2 x 3) when fully supported, and a total score of 8 when partially supported (2 x 4).

I added this information as comments to the code, as suggested.

igorschoester reviewed (changes requested):
CR 🏗 

Please, no `todo` in the code!

As for the discussion(s), I think we could move this to a refactoring issue.
In cooperation with people (Irene!) that know how and why this works the way it does.

I might even accept the magic numbers for now 😉

hansjovis committed:
Added ReadabilityScoreAggregator to assess module using JSDoc statements.

hansjovis committed:
Renamed allAssessmentsSupported to isFullySupported in argument of function because of consistent naming.

hansjovis committed:
Added comments explaining the workings of the penalty and scoring system.

hansjovis committed:
Removed todo comment.

igorschoester commented:
CR & Acceptance 👍
